### PR TITLE
Support for surfacing X.509 client certificates to the Amqp application

### PR DESCRIPTION
**Pull request by rajeevmv**

### CODE REVIEW

File: Microsoft.Azure.Amqp/Amqp/Transport/TlsTransportSetting.cs [hunk 1]
```diff
@@ -62,8 +62,13 @@ public RemoteCertificateValidationCallback CertificateValidationCallback
             get;
             set;
         }
-#endif
 
+        public bool SurfaceRemoteClientCertificate
```
Comment (xinchen10) [inline Microsoft.Azure.Amqp/Amqp/Transport/TlsTransportSetting.cs]: This is not needed. With client cert authentication, the principal should be always X509Principal.

File: Microsoft.Azure.Amqp/Amqp/Transport/TlsTransport.cs [hunk 2]
```diff
@@ -198,10 +199,22 @@ void HandleOpenComplete(IAsyncResult result, bool syncComplete)
                 else
                 {
                     this.sslStream.EndAuthenticateAsServer(result);
-                    if (this.sslStream.RequireMutualAuthentication && this.sslStream.IsRemoteCertificateValid && this.sslStream.RemoteCertificate != null)
+                    if (this.sslStream.RequireMutualAuthentication && this.sslStream.RemoteCertificate != null)
```
Comment (xinchen10) [inline Microsoft.Azure.Amqp/Amqp/Transport/TlsTransport.cs]: If certificate is not valid, identity should not be established. Application can use the default validation logic or provide a validation callback on TlsTransportSettings.

File: Microsoft.Azure.Amqp/Amqp/X509/X509CertificateIdentity.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+        /// <summary>
+        ///  Authentication type accessor
+        /// </summary>
+        public string AuthenticationType { get; }
+
+        /// <summary>
+        ///  IsAuthenticated accessor
+        /// </summary>
+        public bool IsAuthenticated => false;
```
Comment (xinchen10) [inline Microsoft.Azure.Amqp/Amqp/X509/X509CertificateIdentity.cs]: If certificate is valid (accepted), client should be authenticated. Whether this identity has access to resources should be handled by the authorization logic which may perform further checks on the certificate.

File: Microsoft.Azure.Amqp/Amqp/X509/X509CertificateIdentity.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    /// <summary>
+    /// Represents an X509 certificate as a security identity
+    /// </summary>
+    public class X509CertificateIdentity : IIdentity
+    {
+        /// <summary>
+        /// Maximum Clock Skew allowed
+        /// </summary>
+        public static readonly TimeSpan MaxClockSkew = TimeSpan.FromMinutes(5);
```
Comment (xinchen10) [inline Microsoft.Azure.Amqp/Amqp/X509/X509CertificateIdentity.cs]: This should be moved to app code. The library should present the cert as is. Different applications may have different tolerance on clock skew.

File: Microsoft.Azure.Amqp/Amqp/X509/X509CertificateIdentity.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        /// <summary>
+        ///  Actual X509 Certificate
+        /// </summary>
+        public X509Certificate2 Certificate { get; }
+
+        /// <summary>
+        ///  Check is the certificate has expired or is too new
+        /// </summary>
+        /// <returns></returns>
+        public bool IsExpiredOrNotValidYet()
```
Comment (xinchen10) [inline Microsoft.Azure.Amqp/Amqp/X509/X509CertificateIdentity.cs]: This is app logic which can be done either in the remote cert validation callback, or part of the authorization code.

File: Microsoft.Azure.Amqp/Amqp/Transport/TlsTransport.cs (hunk 2 above)
Comment (rajeevmv) [inline Microsoft.Azure.Amqp/Amqp/Transport/TlsTransport.cs]: We want to allow self-signed certs to establish device identity. These certs do not pass certificate validation.

File: Microsoft.Azure.Amqp/Amqp/Transport/TlsTransportSetting.cs (hunk 1 above)
Comment (rajeevmv) [inline Microsoft.Azure.Amqp/Amqp/Transport/TlsTransportSetting.cs]: Removed.

File: Microsoft.Azure.Amqp/Amqp/X509/X509CertificateIdentity.cs (hunk 3 above)
Comment (rajeevmv) [inline Microsoft.Azure.Amqp/Amqp/X509/X509CertificateIdentity.cs]: Addressed.

File: Microsoft.Azure.Amqp/Amqp/X509/X509CertificateIdentity.cs (hunk 4 above)
Comment (rajeevmv) [inline Microsoft.Azure.Amqp/Amqp/X509/X509CertificateIdentity.cs]: Removed.

File: Microsoft.Azure.Amqp/Amqp/X509/X509CertificateIdentity.cs (hunk 5 above)
Comment (rajeevmv) [inline Microsoft.Azure.Amqp/Amqp/X509/X509CertificateIdentity.cs]: Removed.

File: Microsoft.Azure.Amqp/Amqp/Transport/TlsTransport.cs (hunk 2 above)
Comment (rajeevmv) [inline Microsoft.Azure.Amqp/Amqp/Transport/TlsTransport.cs]: Removed the setting to allow self-signed certificates.

**Pull request merged**